pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import os
+
+from pootle.core.url_helpers import split_pootle_path
+from pootle_store.models import Store
+
+
+def split_vfolder_location(path):
+    (lang_code, proj_code,
+     dir_path, filename) = split_pootle_path(path)
+    if lang_code == "{LANG}":
+        lang_code = None
+    if proj_code == "{PROJ}":
+        proj_code = None
+    return lang_code, proj_code, dir_path, filename
+
+
+class VirtualFolderPathMatcher(object):
+
+    def __init__(self, vf):
+        self.vf = vf
+
+    def update_matching_stores(self):
+        existing_stores = self.vf.stores.all()
+        matching_stores = self.find_matching_stores()
+        to_add = []
+        to_delete = []
+        for store in matching_stores:
+            if store not in existing_stores:
+                to_add.append(store)
+        for store in existing_stores:
+            if store not in matching_stores:
+                to_delete.append(store)
+        if to_add:
+            self.vf.stores.add(*to_add)
+        if to_delete:
+            self.vf.stores.delete(*to_delete)
+
+    def add_store_if_matching(self, store):
+        no_add = (
+            not self.path_matches(store.pootle_path)
+            or self.vf.stores.filter(store=store).exists())
+        if no_add:
+            return
+        self.vf.stores.add(store)
+
+    @property
+    def filter_rules(self):
+        return (
+            r.strip()
+            for r
+            in self.vf.filter_rules.split(","))
+
+    @property
+    def path_parts(self):
+        return split_vfolder_location(self.vf.location)
+
+    @property
+    def lang_code(self):
+        return self.path_parts[0]
+
+    @property
+    def proj_code(self):
+        return self.path_parts[1]
+
+    @property
+    def dir_path(self):
+        return self.path_parts[2]
+
+    @property
+    def filename(self):
+        return self.path_parts[3]
+
+    @property
+    def store_qs(self):
+        qs = Store.objects
+        if self.lang_code:
+            qs = qs.filter(translation_project__language__code=self.lang_code)
+        if self.proj_code:
+            qs = qs.filter(translation_project__project__code=self.proj_code)
+        if self.use_regex:
+            if self.lang_code or self.proj_code:
```

Comment:
Thinking it better, this inner `if` is unnecessary. I mean, if either `self.lang_code` or `self.proj_code` are `None` then `self.use_regex` is `True`.